File: go/day01/day01.go
package day01

import (
	"aoc23/inputs"
	"strings"
)

func parse(input string) []string {
	lines := strings.Split(input, "\n")
	lines = lines[:len(lines)-1]
	return lines
}
func Run() [2]int {
	lines := parse(inputs.Input01)
	p1 := part1(lines)
	p2 := part2(lines)

	return [2]int{p1, p2}
}

var numbers = [10]string{
	"zero", "one", "two", "three", "four",
	"five", "six", "seven", "eight", "nine",
}

func part2(lines []string) int {
	var sum int
	for _, line := range lines {
		var num [2]rune
	first:
		for i, r := range line {
			if r >= '0' && r <= '9' {
				num[0] = r - '0'
				break
			}
			for j, n := range numbers {
				if strings.HasSuffix(line[:i+1], n) {
					num[0] = rune(j)
					break first
				}
			}
		}
	last:
		for i := len(line) - 1; i >= 0; i-- {
			r := rune(line[i])
			if r >= '0' && r <= '9' {
				num[1] = r - '0'
				break
			}
			for j, n := range numbers {
				if strings.HasPrefix(line[i:], n) {
					num[1] = rune(j)
					break last
				}
			}
		}
		sum += int(num[0])*10 + int(num[1])
	}
	return sum
}

func part1(lines []string) int {
	var sum int
	for _, line := range lines {
		var num [2]rune
		for _, r := range line {
			if r >= '0' && r <= '9' {
				num[0] = r - '0'
				break
			}
		}
		for i := len(line) - 1; i >= 0; i-- {
			r := rune(line[i])
			if r >= '0' && r <= '9' {
				num[1] = r - '0'
				break
			}
		}
		sum += int(num[0])*10 + int(num[1])
	}
	return sum
}

File: rs/src/day01.rs
pub fn run() -> [u32; 2] {
    let input = include_str!("../inputs/input01.txt");
    let p1 = part01(input);
    let p2 = part02(input);
    [p1, p2]
}

fn part01(input: &str) -> u32 {
    input.lines().map(find_num).sum::<u32>()
}

fn find_num(line: &str) -> u32 {
    let first = line
        .chars()
        .find(|c| c.is_ascii_digit())
        .unwrap()
        .to_digit(10)
        .unwrap();
    let last = line
        .chars()
        .rev()
        .find(|c| c.is_ascii_digit())
        .unwrap()
        .to_digit(10)
        .unwrap();
    first * 10 + last
}

const NUMS: [&str; 10] = [
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn part02(input: &str) -> u32 {
    input.lines().map(find_num2).sum::<u32>()
}

fn find_num2(line: &str) -> u32 {
    let first = line
        .chars()
        .scan("".to_string(), |state, c| {
            if c.is_ascii_digit() {
                return Some(c.to_digit(10));
            }
            state.push(c);
            let num: Option<u32> = NUMS
                .into_iter()
                .enumerate()
                .find(|(_, n)| state.ends_with(n))
                .and_then(|(i, _)| i.try_into().ok());

            Some(num)
        })
        .find(|c| c.is_some())
        .unwrap()
        .unwrap();

    let last = line
        .chars()
        .rev()
        .scan("".to_string(), |state, c| {
            if c.is_ascii_digit() {
                return Some(c.to_digit(10));
            }
            state.push(c);
            let num: Option<u32> = NUMS
                .into_iter()
                .map(|n| n.chars().rev().collect::<String>())
                .enumerate()
                .find(|(_, n)| state.ends_with(n))
                .and_then(|(i, _)| i.try_into().ok());

            Some(num)
        })
        .find(|c| c.is_some())
        .unwrap()
        .unwrap();

    first * 10 + last
}
